crates/next-dev/src/lib.rs

```diff
@@ -192,58 +203,69 @@ impl NextDevServerBuilder {
         let show_all = self.show_all;
         let log_detail = self.log_detail;
         let browserslist_query = self.browserslist_query;
-        let log_options = LogOptions {
+        let log_options = Arc::new(LogOptions {
             current_dir: current_dir().unwrap(),
             show_all,
             log_detail,
             log_level: self.log_level,
-        };
+        });
         let entry_requests = Arc::new(self.entry_requests);
-        let console_ui = Arc::new(ConsoleUi::new(log_options));
-        let console_ui_to_dev_server = console_ui.clone();
         let server_addr = Arc::new(server.addr);
         let tasks = turbo_tasks.clone();
+        let issue_provider = self.issue_reporter.unwrap_or_else(|| {
+            // Initialize a ConsoleUi reporter if no custom reporter was provided
+            Box::new(move || ConsoleUiVc::new(log_options.clone().into()).into())
+        });
+        let issue_reporter_arc = Arc::new(move || issue_provider.get_issue_reporter());
+
+        let get_issue_reporter = issue_reporter_arc.clone();
         let source = move || {
             source(
                 root_dir.clone(),
                 project_dir.clone(),
                 entry_requests.clone().into(),
                 eager_compile,
                 turbo_tasks.clone().into(),
-                console_ui.clone().into(),
+                get_issue_reporter(),
                 browserslist_query.clone(),
                 server_addr.clone().into(),
             )
         };
 
-        Ok(server.serve(tasks, source, console_ui_to_dev_server))
+        Ok(server.serve(tasks, source, issue_reporter_arc.clone()))
     }
 }
 
-async fn handle_issues<T: Into<RawVc>>(source: T, console_ui: ConsoleUiVc) -> Result<()> {
-    let state = console_ui
-        .group_and_display_issues(TransientValue::new(source.into()))
-        .await?;
+async fn handle_issues<T: Into<RawVc> + CollectiblesSource + Copy>(
+    source: T,
+    issue_reporter: IssueReporterVc,
+) -> Result<()> {
+    let issues_vc = IssueVc::peek_issues_with_path(source).await?;
+    let issues = issues_vc.strongly_consistent().await?;
+    issue_reporter.report_issues(
+        TransientInstance::new(issues),
+        TransientValue::new(source.into()),
+    );
 
-    if state.has_fatal {
+    if *issues_vc.has_fatal().await? {
```

Comment:
You probably want to read this strongly_consistent since you are reading issues strongly_consistent.

But there is still no guarantee that issues and fatal are consistent together.

---

So it might be better to use `issues.has_fatal()` instead and about the additional function.